### PR TITLE
[FLINK-16225] Implement user class loading exception handler

### PR DESCRIPTION
**Pull request by azagrebin**
1.10 Port of #12446.
also ports #12563 to reduce netty overhead and direct memory allocation timeout in some tests

### CODE REVIEW

**Comment by azagrebin** [comment]
@flinkbot run travis

**Comment by azagrebin** [comment]
When `EventTimeWindowCheckpointingITCase` runs the whole package, it sometimes hits the default JVM direct memory limit (2g). The GC of netty direct memory buffers is slowed down for some reason which leads to direct OOM errors in netty stack. This is also reproducible in `release-1.10` branch with a lower JVM direct memory limit (e.g. 512mb). This is not clear how inheritance in user class loaders, introduced in this PR, affects direct memory GC.

We could reduce the usage of netty buffers by porting some network stack improvements to 1.10 from 1.11 (FLINK-15962) but we rather do not merge this change to 1.10 and reconsider it if this becomes a problem in future.
cc @zhijiangW

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerSharedServices.java [hunk 1]
```diff
@@ -144,11 +146,13 @@ public static JobManagerSharedServices fromConfiguration(
...
 		final String[] alwaysParentFirstLoaderPatterns = CoreOptions.getParentFirstLoaderPatterns(config);
 
+		final boolean failOnJvmMetaspaceOomError = config.getBoolean(CoreOptions.FAIL_ON_USER_CLASS_LOADING_METASPACE_OOM);
 		final BlobLibraryCacheManager libraryCacheManager =
 			new BlobLibraryCacheManager(
 				blobServer,
 				FlinkUserCodeClassLoaders.ResolveOrder.fromString(classLoaderResolveOrder),
-				alwaysParentFirstLoaderPatterns);
+				alwaysParentFirstLoaderPatterns,
+				fatalErrorHandler);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerSharedServices.java:155]:
```suggestion
				failOnJvmMetaspaceOomError ? fatalErrorHandler : null);
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 2]
```diff
@@ -1326,7 +1326,8 @@ private JobManagerConnection associateWithJobManager(
 		final LibraryCacheManager libraryCacheManager = new BlobLibraryCacheManager(
 			blobCacheService.getPermanentBlobService(),
 			taskManagerConfiguration.getClassLoaderResolveOrder(),
-			taskManagerConfiguration.getAlwaysParentFirstLoaderPatterns());
+			taskManagerConfiguration.getAlwaysParentFirstLoaderPatterns(),
+			fatalErrorHandler);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java:1330]: Here we are not respecting `FAIL_ON_USER_CLASS_LOADING_METASPACE_OOM`.

**Comment by azagrebin** [comment]
Thanks for the review @tillrohrmann 
True, the option got lost during the rebase.

**Pull request merged**